File: src/javascript/ssrf.js
// CWE-918: Server-Side Request Forgery (SSRF).

const express = require('express');
const axios = require('axios');

const app = express();

app.get('/fetch', async (req, res) => {
  const url = req.query.url;
  const r = await axios.get(url);
  res.type('text/plain').send(r.data);
});

app.get('/preview', async (req, res) => {
  const target = req.query.target;
  const html = (await axios.get('http://' + target)).data;
  res.type('text/html').send(html);
});

app.get('/avatar', async (req, res) => {
  const user = req.query.user;
  const r = await axios.get(`http://internal-avatars.local/${user}.png`, { responseType: 'arraybuffer' });
  res.type('image/png').send(Buffer.from(r.data));
});

app.listen(3004);

File: src/python/ssrf.py
"""CWE-918: SSRF via requests on user-controlled URLs."""

import requests
from flask import Flask, request

app = Flask(__name__)


@app.route("/fetch")
def fetch():
    url = request.args.get("url", "")
    r = requests.get(url, timeout=5)
    return r.text


@app.route("/preview")
def preview():
    target = request.args.get("target", "")
    return requests.get("http://" + target).text


@app.route("/webhook", methods=["POST"])
def webhook():
    callback = request.json.get("callback")
    requests.post(callback, json={"event": "ping"})
    return "ok"


if __name__ == "__main__":
    app.run(port=4004)
